src/compiler/sys.ts

```diff
@@ -294,6 +294,50 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const watchedDirectories: { [path: string]: FileWatcher } = {};
+                const watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    const watchedPaths = Object.keys(watchedDirectories);
+                    // Try to find parent paths that are already watched. If found, don't add directory watchers
+                    const watchedParentPaths = watchedPaths.filter(path => fileName.indexOf(path) === 0);
+                    // If adding new watchers, try to find children paths that are already watched. If found, close them. 
+                    if (watchedParentPaths.length === 0) {
+                        const pathToWatch = ts.getDirectoryPath(fileName);
+                        for (const watchedPath in watchedDirectories) {

```

Comment:
I would think twice before completely switching to recursive watchers or at least test it in different scenarios. Effectively it means that if user adds some file that is close to the root of the disk then watcher will receive tons of notifications about files he doesn't care